commit d3aabb9325aed958e157af6c3b2be371d36e3333
Author: Jeff Law <jlaw@ventanamicro.com>
Date:   Mon Dec 4 10:12:42 2023 -0700

    Revert "tree-optimization/112827 - corrupt SCEV cache during SCCP"
    
    This reverts commit de0ab339a795352c843f6e9b2dfce222f26588de.

diff --git a/gcc/testsuite/gcc.dg/torture/pr112827-1.c b/gcc/testsuite/gcc.dg/torture/pr112827-1.c
deleted file mode 100644
index 6838cbbe62f..00000000000
--- a/gcc/testsuite/gcc.dg/torture/pr112827-1.c
+++ /dev/null
@@ -1,14 +0,0 @@
-/* { dg-do compile } */
-
-int a, b, c, d, e;
-int main() {
-  for (; c; c++) {
-    for (a = 0; a < 2; a++)
-      ;
-    for (; b; b++) {
-      e = d;
-      d = a;
-    }
-  }
-  return 0;
-}
diff --git a/gcc/testsuite/gcc.dg/torture/pr112827-2.c b/gcc/testsuite/gcc.dg/torture/pr112827-2.c
deleted file mode 100644
index a7a2a70211b..00000000000
--- a/gcc/testsuite/gcc.dg/torture/pr112827-2.c
+++ /dev/null
@@ -1,18 +0,0 @@
-/* { dg-do compile } */
-
-short a, b[1], f;
-char c, g;
-int d, e;
-int main() {
-  for (; f; f++) {
-    for (d = 0; d < 2; d++)
-      ;
-    if (a)
-      for (g = 0; g < 2; g++)
-        for (c = 0; c < 2; c += b[d+g])
-          ;
-    for (; e; e++)
-      ;
-  }
-  return 0;
-}
diff --git a/gcc/tree-scalar-evolution.cc b/gcc/tree-scalar-evolution.cc
index 7556d89e9f8..065bcd0743d 100644
--- a/gcc/tree-scalar-evolution.cc
+++ b/gcc/tree-scalar-evolution.cc
@@ -3847,10 +3847,13 @@ final_value_replacement_loop (class loop *loop)
       def = unshare_expr (def);
       remove_phi_node (&psi, false);
 
-      /* Propagate constants immediately, but leave an unused initialization
-	 around to avoid invalidating the SCEV cache.  */
+      /* Propagate constants immediately.  */
       if (CONSTANT_CLASS_P (def))
-	replace_uses_by (rslt, def);
+	{
+	  replace_uses_by (rslt, def);
+	  release_ssa_name (rslt);
+	  continue;
+	}
 
       /* Create the replacement statements.  */
       gimple_seq stmts;
